cores/esp8266/WString.h

```diff
@@ -284,17 +284,17 @@ class String {
             uint16_t len;
         };
         // This allows strings up up to 11 (10 + \0 termination) without any extra space.
-        enum { SSOSIZE = sizeof(struct _ptr) + 4 - 1 }; // Characters to allocate space for SSO, must be 12 or more
+        static constexpr auto SSOSIZE = sizeof(struct _ptr) + 4 - 1; // Characters to allocate space for SSO, must be 12 or more
         struct _sso {
             char     buff[SSOSIZE];
             unsigned char len   : 7; // Ensure only one byte is allocated by GCC for the bitfields
             unsigned char isSSO : 1;
         } __attribute__((packed)); // Ensure that GCC doesn't expand the flag byte to a 32-bit word for alignment issues
-        enum { CAPACITY_MAX = 65535 }; // If typeof(cap) changed from uint16_t, be sure to update this enum to the max value storable in the type
         union {
             struct _ptr ptr;
             struct _sso sso;
         };
+        static constexpr auto CAPACITY_MAX = (1U << sizeof(ptr.cap) * 8) - 1;
```

Comment:
This is not used in the .h file, so why not move it to the .cpp?

Reply:
> (it is protected, so only those inheriting from this class)

just only, but still possible.
if so, we should warn: [API change] (of course `.isNotSSO` too)

i'd like to know why these internal implementation members are protected but not private.

Reply:
Well the `init()` function was really needed for proper functioning of the class, so if someone wants to make a derived class, this function must be protected (or public).
I can imagine it is not needed for using the String class, so then protected makes perfect sense.
Now init() is no longer needed in the current implementation, but it may indeed break API design if the function is removed. The same for the `CAPACITY_MAX` value.